samples/order/frontend.go

```diff
@@ -12,19 +12,193 @@
-// TODO(jba): implement
-
+// A frontend is a web server that takes image-processing orders.
 type frontend struct {
 	requestTopic *pubsub.Topic
 	responseSub  *pubsub.Subscription
 	bucket       *blob.Bucket
 	coll         *docstore.Collection
 }
+
+var (
+	listTemplate      = template.Must(template.ParseFiles("list.htmlt"))
+	orderFormTemplate = template.Must(template.ParseFiles("order-form.htmlt"))
+)
+
+// run starts the server on port and runs it indefinitely.
+func (f *frontend) run(ctx context.Context, port int) error {
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) { http.ServeFile(w, r, "index.html") })
+	http.HandleFunc("/orders/", wrap(f.listOrders))
+	http.HandleFunc("/orders/new", wrap(f.orderForm))
+	http.HandleFunc("/createOrder", wrap(f.createOrder))
+
+	rl := requestlog.NewNCSALogger(os.Stdout, func(err error) { fmt.Fprintf(os.Stderr, "%v\n", err) })
+	s := server.New(nil, &server.Options{
+		RequestLogger: rl,
+	})
+	return s.ListenAndServe(fmt.Sprintf(":%d", port))
+}
+
+// wrap turns handlers that return error into ordinary http.Handlers, by calling http.Error on non-nil
+// errors.
+func wrap(f func(http.ResponseWriter, *http.Request) error) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if err := f(w, r); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	}
+}
+
+// orderForm serves a page that lets the user input a new order.
+func (*frontend) orderForm(w http.ResponseWriter, r *http.Request) error {
+	if r.Method != "GET" {
+		http.Error(w, "bad method for orderForm: want GET", http.StatusBadRequest)
+		return nil
+	}
+	return executeTemplate(orderFormTemplate, nil, w)
+}
+
+// createOrder handles a submitted order form.
+func (f *frontend) createOrder(w http.ResponseWriter, r *http.Request) error {
+	if r.Method != "POST" {
+		http.Error(w, "bad method for createOrder: want POST", http.StatusBadRequest)
+		return nil
+	}
+	email := r.FormValue("email")
+	if email == "" {
+		http.Error(w, "email missing", http.StatusBadRequest)
+		return nil
+	}
+
+	file, _, err := r.FormFile("file")
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if _, err := f.doCreateOrder(r.Context(), email, file, time.Now()); err != nil {
+		return err
+	}
+	fmt.Fprintln(w, "Order received. Thank you.")
+	return nil
+}
+
+// doCreateOrder returns the order ID it generates, for testing.
+func (f *frontend) doCreateOrder(ctx context.Context, email string, file io.Reader, now time.Time) (id string, err error) {
+	// Assign an ID for the order here, rather than in the processor.
+	// That allows the processor to detect duplicate pub/sub messages.
+	id = f.newID()
+	req := &OrderRequest{
+		ID:         id,
+		InImage:    id + "-in",
+		Email:      email,
+		CreateTime: now,
+	}
+
+	// Copy the uploaded input file to the bucket.
+	w, err := f.bucket.NewWriter(ctx, req.InImage, nil)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
```

Comment:
This is a bit hard to follow because the flow of control is inverted :-/

Reply:
Rewrote.